### PR TITLE
Feat: Add option one to one relation in Entity Field

### PR DESCRIPTION
**Pull request by morhag90**
Close: #1215

## PR Details

Add the option to choose the side for the FK in a one-to-one relation field. 

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

### CODE REVIEW

File: libs/data-service-generator/src/prepare-context.ts [hunk 1]
```diff
@@ -212,8 +212,9 @@ function resolveLookupFields(entities: Entity[]): Entity[] {
           //**@todo: in one-to-one relation, only one side should have a foreign key.
           //We currently decide randomly based on sorting the permanent ID
           //instead we should let the user decide which side holds the foreign key  */
+          //
           const isOneToOneWithoutForeignKey =
-            isOneToOne && field.permanentId > relatedField.permanentId;
+            isOneToOne && field.permanentId !== fieldProperties.FkHolder;
```
Comment (yuval-hazaz) [inline libs/data-service-generator/src/prepare-context.ts:217]: you should keep the current behavior in case the new property is undefined or null

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      !relatedField.properties.allowMultipleSelection
+    );
+  }, [data, formik.values]);
+
+  const entityListOptions = useMemo(() => {
+    return (
+      (entity &&
+        data && [
+          {
+            value: currentEntity.permanentId,
```
Comment (yuval-hazaz) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:49]: This is not correct, you should use the value from formik with the current field

File: packages/amplication-prisma-db/.env [hunk 3]
```diff
@@ -4,3 +4,4 @@ POSTGRESQL_URL=postgresql://admin:admin@localhost:5432/${SERVICE_DB_NAME}
 POSTGRESQL_USER=admin
 POSTGRESQL_PASSWORD=admin
 SERVICE_DB_NAME=amplication
+PRISMA_CLIENT_OUTPUT_PATH= generated-prisma-client
```
Comment (yuval-hazaz) [inline packages/amplication-prisma-db/.env:7]: is that related to this PR?

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-prisma-db/.env (hunk 3 above)
Comment (morhag90) [reply packages/amplication-prisma-db/.env:7]: @yuval-hazaz  No, it's a new env var that is missing in the env file

File: libs/data-service-generator/src/prepare-context.ts [hunk 4]
```diff
@@ -209,11 +209,9 @@ function resolveLookupFields(entities: Entity[]): Entity[] {
             !fieldProperties.allowMultipleSelection &&
             !relatedFieldProperties.allowMultipleSelection;
 
-          //**@todo: in one-to-one relation, only one side should have a foreign key.
-          //We currently decide randomly based on sorting the permanent ID
-          //instead we should let the user decide which side holds the foreign key  */
           const isOneToOneWithoutForeignKey =
-            isOneToOne && field.permanentId > relatedField.permanentId;
+            (isOneToOne && field.permanentId > relatedField.permanentId) ||
```
Comment (yuval-hazaz) [inline libs/data-service-generator/src/prepare-context.ts:213]: We should first check if the new property is populated and only use the previous condition in case it is not

File: packages/amplication-prisma-db/.env (hunk 3 above)
Comment (yuval-hazaz) [reply packages/amplication-prisma-db/.env:7]: why not merging the missing variable into this branch?

File: packages/amplication-server/src/core/entity/entity.service.ts [hunk 5]
```diff
@@ -2026,7 +2028,9 @@ export class EntityService {
     entityId: string,
     relatedEntityId: string,
     relatedFieldId: string,
-    user: User
+    user: User,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    FkHolder?: string
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/entity/entity.service.ts:2033]: rename to fkHolder

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
@morhag90 in addition to what I commented inline, I have doubts about the way we delete and recreate the related field - it is not part of this PR, but I think we should look at it anyway

File: packages/amplication-prisma-db/.env (hunk 3 above)
Comment (morhag90) [reply packages/amplication-prisma-db/.env:7]: @yuval-hazaz, that's what I did, no??....

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import { GET_ENTITY_FIELD_BY_PERMANENT_ID } from "./RelatedEntityFieldField";
+
+type Props = Omit<SelectFieldProps, "options"> & {
+  entity: models.Entity;
+};
+
+const RelationFkHolderField = ({ entity, ...props }: Props) => {
+  const formik = useFormikContext<models.EntityField>();
+
+  const { data } = useQuery<{ entity: models.Entity }>(
```
Comment (barshimi) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:15]: move it to query file and also add `skip` so it wont run every time.

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    GET_ENTITY_FIELD_BY_PERMANENT_ID,
+    {
+      variables: {
+        entityId: formik.values.properties.relatedEntityId,
+        fieldPermanentId: formik.values.properties.relatedFieldId,
+      },
+    }
+  );
+
+  const relatedField =
```
Comment (barshimi) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:25]: add state for that and listen to changes in data

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    data &&
+    data.entity &&
+    data.entity.fields &&
+    data.entity.fields.length &&
+    data.entity.fields[0];
+
+  const isOneToOne = useMemo(() => {
```
Comment (barshimi) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:32]: do you want to memo the result of this ? if not it should also be part of the state

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    return (
+      entity &&
+      data &&
+      !formik.values?.properties?.allowMultipleSelection &&
+      !relatedField.properties?.allowMultipleSelection
+    );
+  }, [data, formik.values]);
+
+  const entityListOptions = useMemo(() => {
```
Comment (barshimi) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:41]: this is also problematic. this function should get the data as argument

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx (hunk 7 above)
Comment (yuval-hazaz) [reply packages/amplication-client/src/Entity/RelationFkHolderField.tsx:25]: we better use useMemo instead of state

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx (hunk 8 above)
Comment (yuval-hazaz) [reply packages/amplication-client/src/Entity/RelationFkHolderField.tsx:32]: useMemo is better in this case

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx (hunk 9 above)
Comment (yuval-hazaz) [reply packages/amplication-client/src/Entity/RelationFkHolderField.tsx:41]: why?

File: packages/amplication-prisma-db/.env (hunk 3 above)
Comment (yuval-hazaz) [reply packages/amplication-prisma-db/.env:7]: please use a different PR

File: packages/amplication-server/src/core/entity/entity.service.ts [hunk 10]
```diff
@@ -2175,13 +2182,40 @@ export class EntityService {
...
+        )?.fkHolder;
+
+        await this.prisma.entityField.update({
+          where: {
+            id: relatedField.id,
+          },
+          data: {
+            properties: relatedFieldProps as unknown as Prisma.InputJsonValue,
+          },
+        });
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/entity/entity.service.ts:2216]: only if (field.dataType === EnumDataType.Lookup && properties.fkHolder !== null)

File: libs/data-service-generator/src/prepare-context.ts [hunk 11]
```diff
@@ -209,11 +209,9 @@ function resolveLookupFields(entities: Entity[]): Entity[] {
             !fieldProperties.allowMultipleSelection &&
             !relatedFieldProperties.allowMultipleSelection;
 
-          //**@todo: in one-to-one relation, only one side should have a foreign key.
-          //We currently decide randomly based on sorting the permanent ID
-          //instead we should let the user decide which side holds the foreign key  */
           const isOneToOneWithoutForeignKey =
-            isOneToOne && field.permanentId > relatedField.permanentId;
+            (isOneToOne && field.permanentId !== fieldProperties.fkHolder) ||
+            (isOneToOne && field.permanentId > relatedField.permanentId);
```
Comment (yuval-hazaz) [inline libs/data-service-generator/src/prepare-context.ts:214]: separate conditions for when fieldProperties.fkHolder is null (or undefined) and not null

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 12]
```diff
@@ -33,8 +33,8 @@ const RelationFkHolderField = ({ entity, ...props }: Props) => {
     return (
       entity &&
       data &&
-      !formik.values?.properties.allowMultipleSelection &&
-      !relatedField.properties.allowMultipleSelection
+      !formik.values?.properties?.allowMultipleSelection &&
+      !relatedField.properties?.allowMultipleSelection
```
Comment (abrl91) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:37]: Why do we now need the `?` after `properties`?

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    if (!data) {
+      entityFieldRef.current = {
+        relatedField: undefined,
+        isOneToOne: false,
+        entityListOptions: [],
+      };
+      return;
+    }
+
+    console.log("RelationFkHolderField", formik.values);
```
Comment (yuval-hazaz) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:47]: remove console

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        entityId: formik.values.properties.relatedEntityId,
+        fieldPermanentId: formik.values.properties.relatedFieldId,
+      },
+      skip:
+        !formik.values.properties?.relatedEntityId ||
+        !formik.values.properties?.relatedFieldId,
+    }
+  );
+
+  useEffect(() => {
```
Comment (yuval-hazaz) [inline packages/amplication-client/src/Entity/RelationFkHolderField.tsx:37]: I understand what you did here, but I think the useMemo version is much more readable and straight-forward - don't see the advantage of using the ref object. Also, useMemo is consistent with the way we do it across the client app
I don't ask you to change, but I just want to have a discussion about the best practice

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-client/src/Entity/RelationFkHolderField.tsx (hunk 12 above)
Comment (morhag90) [reply packages/amplication-client/src/Entity/RelationFkHolderField.tsx]: @abrl91 because they init will null values, and this component is triggered.

**Review by yuval-hazaz — APPROVED** [review]
